src/data/resolvers/colonyActions.ts

```diff
@@ -0,0 +1,180 @@
+       */
+      const transactionReceipt = await provider.getTransactionReceipt(
+        transactionHash,
+      );
+
+      if (transactionReceipt) {
+        const {
+          transactionHash: hash,
+          status,
+          logs,
+          blockHash,
+          from,
+        } = transactionReceipt;
+
+        /*
+         * Get the block time in ms
+         *
+         * we fallback to 0, which is 1/1/1970 :)
+         * If we don't find a time for the current tx (which shouldn't happen actually)
+         */
+        const createdAt = blockHash
+          ? await getBlockTime(provider, blockHash)
+          : 0;
+
+        /*
+         * @NOTE Parse all logs with all clients to generate all the possible events
+         * This is the second iteration of this implementation.
+         *
+         * It is less perfomant than just iterating over the clientsInstancesArray and
+         * then just passing it the logs array to process, but this way allows us to
+         * preseve the position of each event as it was originally emmited by the contracts.
+         *
+         * This way we can use that index position to inferr the action type from them
+         */
+        const reverseSortedEvents = logs
+          ?.map((log) => {
+            const [parsedLog] = clientsInstancesArray
+              .map((clientType) => {
+                const type = clientType?.clientType;
+                const potentialParsedLog = clientType?.interface.parseLog(log);
+                if (potentialParsedLog) {
+                  const { name, values } = potentialParsedLog;
+                  return {
+                    name,
+                    values,
+                    createdAt,
+                    emmitedBy: type,
+                  };
+                }
+                return null;
+              })
+              .filter((potentialLog) => !!potentialLog);
+            return parsedLog;
+          })
+          .reverse();
+
+        const payment = {
+          recipient: AddressZero,
+          fromDomain: 1,
+          amount: '0',
+          tokenAddress: AddressZero,
+        };
+        let paymentDetails;
+
+        const actionType = getActionType(reverseSortedEvents);
+
+        if (actionType === ColonyActions.Payment) {
+          /*
+           * If the action is a `Payment` type event, it will most definetly have
+           * these two events already, so we don't need to bother checking for
+           * their existance
+           */
+          const paymentAddedEvent = reverseSortedEvents?.find(
+            (event) => event?.name === ColonyAndExtensionsEvents.PaymentAdded,
+          );
+          const payoutClaimedEvent = reverseSortedEvents?.find(
+            (event) => event?.name === ColonyAndExtensionsEvents.PayoutClaimed,
+          );
+          const { paymentId } = paymentAddedEvent?.values;
+          paymentDetails = await getPaymentDetails(
+            paymentId,
```

Comment:
what if this will be undefined?

Reply:
It actually cannot be `undefined`. If that happens _somehow_ it means the contracts are actually broken, and the app crashes long before that :smile: 

Also, let's just consider that it is `undefined`, and the app didn't crash so far. In that case, `getPaymentDetails` will return `null`. So not that big of a deal.

I will however put a check for if `paymentDetails` exists, before trying to assign values from it.